python/pyiceberg/table/__init__.py

```diff
@@ -934,206 +963,611 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
+                )
+
+        if has_changes:
             return StructType(*new_fields)
 
         return struct
 
-    def field(self, field: NestedField, field_result: IcebergType) -> IcebergType:
-        field_id: int = field.field_id
-        if field_id in self.adds:
-            new_fields = self.adds[field_id]
-            if len(new_fields) > 0:
-                fields = _ApplyChanges.add_fields(field_result.fields, new_fields)
-                if len(fields) > 0:
-                    return StructType(*fields)
+    def field(self, field: NestedField, field_result: Optional[IcebergType]) -> Optional[IcebergType]:
+        # the API validates deletes, updates, and additions don't conflict handle deletes
+        if field.field_id in self._deletes:
+            return None
 
-        return field_result
+        # handle updates
+        if (update := self._updates.get(field.field_id)) and field.field_type != update.field_type:
+            return update.field_type
 
-    def list(self, list_type: ListType, element_result: IcebergType) -> IcebergType:
-        element_field: NestedField = list_type.element_field
-        element_type = self.field(element_field, element_result)
-        if element_type is None:
-            raise ValueError(f"Cannot delete element type from list: {element_field}")
+        # handle add & moves
+        added = self._adds.get(field.field_id)
+        moves = self._moves.get(field.field_id)
+        if added is not None or moves is not None:
+            if not isinstance(field.field_type, StructType):
+                raise ValueError(f"Cannot add fields to non-struct: {field}")
 
-        is_element_optional = not list_type.element_required
+            if new_fields := _add_and_move_fields(field.field_type.fields, added or [], moves or []):
+                return StructType(*new_fields)
 
-        if is_element_optional == element_field.required and list_type.element_type == element_type:
-            return list_type
+        return field_result
+
+    def list(self, list_type: ListType, element_result: Optional[IcebergType]) -> Optional[IcebergType]:
+        element_type = self.field(list_type.element_field, element_result)
+        if element_type is None:
+            raise ValueError(f"Cannot delete element type from list: {element_result}")
 
-        return ListType(list_type.element_id, element_type, is_element_optional)
+        return ListType(element_id=list_type.element_id, element=element_type, element_required=list_type.element_required)
 
-    def map(self, map_type: MapType, key_result: IcebergType, value_result: IcebergType) -> IcebergType:
+    def map(
+        self, map_type: MapType, key_result: Optional[IcebergType], value_result: Optional[IcebergType]
+    ) -> Optional[IcebergType]:
         key_id: int = map_type.key_field.field_id
-        if key_id in self.adds:
+        if key_id in self._adds:
```

Comment:
I think there are a lot more checks here in the Java version.

Reply:
Yes, added these as well